### PR TITLE
New Testing Infra

### PR DESCRIPTION
**Pull request by karya0**
This PR adds a modern DMTCP testing foundation around a cleaner autotest.py harness, structured dmtcp_command --json output, coordinator protocol coverage, and optional parallel integration execution.

Key changes:

  - Adds typed coordinator command/status JSON output for dmtcp_command.
  - Introduces a reusable dmtcp::Json helper for internal JSON serialization.
  - Replaces the old monolithic autotest runner with a single organized Python harness while preserving autotest_old.py for reference.
  - Adds integrated suites for unit, harness, coordinator protocol, and checkpoint/restart integration tests.
  - Adds autotest.py --jobs N to run integration tests in parallel while keeping output readable.
  - Adds C++ unit test infrastructure for internal message/header coverage.
  - Adds synthetic coordinator worker tests and real-worker JSON control coverage.
  - Updates make check and CI to run the unified harness.

Verification
```zmake -j4
  make -C test check-unit
  python3 test/test_dmtcp_command_json.py
  python3 test/autotest_test.py
  make -C test check-harness-unit
  ./test/autotest.py --suite all command-json-bcheckpoint
  ./test/autotest.py --suite all --jobs 4 command-json-bcheckpoint
  ./test/autotest.py --jobs 4 dmtcp1 dmtcp2 gettid file1
  ./test/autotest.py --jobs 8 dmtcp1 dmtcp2 gettid file1 file3 stat mmap1 poll
  ```

## Summary by CodeRabbit

* **New Features**
  * Command-line tool now emits machine-readable JSON for help/version/status/commands and coordinator responses.

* **Bug Fixes**
  * Improved coordinator handshake, validation, and clearer restart-peer-mismatch reporting; stronger handling of barrier/restart edge cases.

* **Tests**
  * Large new unit and integration test suites, synthetic-worker integration tests, and a new Python test runner covering JSON and coordinator behaviors.

* **Documentation**
  * Added test-harness migration and coordinator coverage guides.

* **Chores**
  * Reworked build/test targets, CI check invocation, and updated ignore rules.

### CODE REVIEW

File: test/unit/main.cpp [hunk 1]
```diff
@@ -0,0 +1,23 @@
+#include "unit_test.h"
+
+#include <array>
+#include <cstddef>
+#include <span>
```
[the remark this answers is not in the document]
Comment (karya0) [reply test/unit/main.cpp:5]: Verified against the current branch: this finding is stale. The branch is based on main commit 171d280b, which adds `AX_CXX_COMPILE_STDCXX([20], [noext], [mandatory])` in `configure.ac`; configured builds compile C++ with `g++ -std=c++20`, and the test/unit targets inherit that configured C++ mode. No extra per-test standard flag is needed. Local validation passed with `make -j4`, direct command/synthetic/harness suites, and `make check AUTOTEST="--jobs 8"`.

File: Makefile.in [hunk 2]
```diff
@@ -222,6 +232,15 @@ check-32: tests-32
 	   ' python to path or build Python 2 or use python3.'; \
 	  fi
 
+check-unit:
+	cd test && $(MAKE) check-unit
```
[the remark this answers is not in the document]
Comment (karya0) [reply Makefile.in]: Fixed in current head d0021169. The top-level `check-unit` target now depends on `build`, so clean-tree invocations build the required archives before descending into `test`. Validated with `make check-unit` and the full `make check AUTOTEST="--jobs 8"` run.

File: src/dmtcp_command.cpp [hunk 3]
```diff
@@ -111,91 +219,106 @@ main(int argc, char **argv)
...
-  case 'k':
-  case 'q':
+  case DMT_KILL:
+  case DMT_QUIT:
     workerList =
-      CoordinatorAPI::connectAndSendUserCommand(cmdChar, &coordCmdStatus);
+      CoordinatorAPI::connectAndSendUserCommand(command, &coordCmdStatus,
+                                                NULL, NULL, NULL,
+                                                &coordResponse);
     break;
```
[the remark this answers is not in the document]
Comment (karya0) [reply src/dmtcp_command.cpp:285]: Fixed in current head d0021169. `dmtcp_command` now uses a shared `freeWorkerList()` cleanup path and releases `workerList` in the JSON error path, JSON success path, non-JSON list/status path, and final non-JSON fallback path. Covered by the command JSON tests and full `make check AUTOTEST="--jobs 8"`.

File: test/autotest_old.py [hunk 4]
```diff
@@ -0,0 +1,1314 @@
...
+
+#get testconfig
+# This assumes Makefile.in in main dir, but only Makefile in test dir.
+try:
+  sys.path += [os.getenv("PWD") + '/test', os.getenv('PWD')]
+  from autotest_config import *
+
+except ImportError:
+  print("\n*** Error importing autotest_config.py: ")
+  sys.exit()
```
[the remark this answers is not in the document]
Comment (karya0) [reply test/autotest_old.py]: Fixed in current head d0021169. `autotest_old.py` now exits with status 1 on `autotest_config.py` import failure, so setup failures are visible to callers instead of looking successful.

File: test/coordinator_synthetic_worker.cpp [hunk 5]
```diff
@@ -0,0 +1,547 @@
...
+    throw std::runtime_error("invalid port");
+  }
+  return static_cast<int>(port);
+}
+
+int
+parsePositiveInt(const char *text)
+{
+  return parsePort(text);
+}
```
[the remark this answers is not in the document]
Comment (karya0) [reply test/coordinator_synthetic_worker.cpp:161]: Fixed in current head d0021169. `parsePositiveInt()` now parses and validates generic positive integers up to `std::numeric_limits<int>::max()` instead of delegating to port-range validation. Added a synthetic-worker regression test using `--hold-seconds 70000` to cover this.

File: test/autotest_old.py [hunk 6]
```diff
@@ -0,0 +1,1314 @@
...
+          x.stderr.close()
+      except:
+        None
+      try:
+        os.waitpid(x.pid, os.WNOHANG)
+      except OSError as e:
+        if e.errno != errno.ECHILD:
+          raise e
+      procs.remove(x)
+
```
[the remark this answers is not in the document]
Comment (karya0) [reply test/autotest_old.py]: Fixed in current head d0021169. The legacy cleanup loop now iterates over `list(procs)` so removing processed entries cannot skip remaining processes/fds.

File: test/unit/unit_test.h [hunk 7]
```diff
@@ -0,0 +1,95 @@
+#ifndef DMTCP_TEST_UNIT_TEST_H
+#define DMTCP_TEST_UNIT_TEST_H
+
+#include <exception>
+#include <iostream>
+#include <sstream>
+#include <span>
+#include <string>
+
```
[the remark this answers is not in the document]
Comment (karya0) [reply test/unit/unit_test.h:10]: Fixed in current head d0021169. `test/unit/unit_test.h` now includes `<utility>` directly for `std::move`, removing the transitive-include dependency.

File: src/dmtcp_coordinator.cpp [hunk 8]
```diff
@@ -1780,16 +1936,20 @@ main(int argc, char **argv)
...
+      JASSERT(dup2(nullFd, STDIN_FILENO) == STDIN_FILENO) (JASSERT_ERRNO)
+        .Text("failed to redirect daemon stdin to /dev/null");
       close(nullFd);
     }
-    JASSERT(dup2(fd, STDOUT_FILENO) == STDOUT_FILENO);
-    JASSERT(dup2(fd, STDERR_FILENO) == STDERR_FILENO);
+    JASSERT(dup2(fd, STDOUT_FILENO) == STDOUT_FILENO) (JASSERT_ERRNO)
+      .Text("failed to redirect daemon stdout");
+    JASSERT(dup2(fd, STDERR_FILENO) == STDERR_FILENO) (JASSERT_ERRNO)
+      .Text("failed to redirect daemon stderr");
```
[the remark this answers is not in the document]
Comment (karya0) [reply src/dmtcp_coordinator.cpp:1897]: Fixed in current head d0021169. Coordinator daemon log creation now uses mode `0644` instead of `0666`, so newly created log files are not world-writable. Covered by rebuild and full `make check AUTOTEST="--jobs 8"` validation.

File: test/Makefile.in [hunk 9]
```diff
@@ -78,10 +78,61 @@ endif
...
+
+check-unit: $(UNIT_TESTS)
+	python3 $(srcdir)/autotest.py --suite unit
+
+check-harness-unit:
+	python3 $(srcdir)/autotest.py --suite harness
+
+check-coordinator-synthetic: $(COORDINATOR_SYNTHETIC_WORKER)
+	python3 $(srcdir)/autotest.py --suite coordinator
+
```
[the remark this answers is not in the document]
Comment (karya0) [reply test/Makefile.in:117]: Fixed in current head d0021169. `test/Makefile.in` now defines a `PYTHON` fallback from the configured Python probes and uses it for `check-unit`, `check-harness-unit`, and `check-coordinator-synthetic` instead of hardcoding `python3`.

File: test/autotest_old.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (karya0) [reply test/autotest_old.py]: Obsolete code.

File: test/autotest_old.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (karya0) [reply test/autotest_old.py]: Obsolete code.

File: test/coordinator_synthetic.py [hunk 10]
```diff
@@ -0,0 +1,1043 @@
...
+
+    def run_command(self, *args):
+        return subprocess.run(
+            [str(DMTCP_COMMAND), *args],
+            cwd=str(ROOT),
+            text=True,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            check=False,
+        )
```
[the remark this answers is not in the document]
Comment (karya0) [reply test/coordinator_synthetic.py:418]: Minor.

**Review by gc00 — APPROVED** [review]
The PR is a good one.  But I have a set of comments, based also on me testing the new autotest on WSL, using Ubuntu 24.04.  I use a horizontal line to separate my comments here.  (And I might have had a small number of inline comments, or maybe none.)

First, two of the coderabbit nitpicks seem useful:

----

src/Makefile.in (1)
402-403: ⚡ Quick win

json.Po is listed for regeneration but not included in dep includes.

am__depfiles_remade now includes ./$(DEPDIR)/json.Po, but there is no matching @am__include .../json.Po include marker later in the file. This can make incremental rebuilds miss header-triggered recompiles for json.cpp.

----

**NOTE:  I'm suggesting to use 0644, _not_ 0640.  Let others read, but not write.**
src/dmtcp_coordinator.cpp (1)
1941-1942: 💤 Low value

Consider restricting log file permissions.

The log file is created with mode 0666, which (before umask) allows world-readable and world-writable access. While umask typically reduces this, explicitly using more restrictive permissions like 0640 or 0644 is safer.

🔒 Suggested fix
-      fd = open(flags.logFilename.c_str(), O_CREAT | O_WRONLY | O_APPEND, 0666);
+      fd = open(flags.logFilename.c_str(), O_CREAT | O_WRONLY | O_APPEND, 0644);

----

The `vim` test seems to be missing.

----

It used to be that `make check-python` would list all the tests, even though it ran only `python`.  We need some way to list all the tests without running them.  Maybe `make check-list` (`python3 autotest.py list`) ??

----

`make check 2>&1 | cat - > tmp.out`  is hanging when it reaches:  `Coordinator protocol tests`
Doing `make check` in background should be useful.  Not sure why it's hanging.

----

The test for command-json is failing on my Windows/WSL, running `Ubuntu 24.04.1 LTS`

  command-json                FAIL. exit=1 Ran 14 tests in 3.613s
    command: /usr/bin/python3 ./test_dmtcp_command_json.py
    cwd: /home/gene/dmtcp.git/test
    output tail (last 13 lines):
      ......F.......
      ======================================================================
      FAIL: test_json_escapes_coordinator_host (__main__.DmtcpCommandJsonTest.test_json_escapes_coordinator_host)
      ----------------------------------------------------------------------
      Traceback (most recent call last):
        File "/home/gene/dmtcp.git/test/./test_dmtcp_command_json.py", line 104, in test_json_escapes_coordinator_host
          self.assertEqual(result.returncode, 2, result.stderr)
      AssertionError: -11 != 2 :

      ----------------------------------------------------------------------
      Ran 14 tests in 3.613s

      FAILED (failures=1)

----

Just a thought.  Should we add a commit to change `DMT_` to `DMTCP_` everywhere?  We had `DMT_` from the earliest days.  But unusual abbreviations can be confusing.
You can decide whether to add such a commit.

----

In `make check`, the display is going for more than 80 characters.  (Although, `make check-alarm` does not go beyond 80 characters.)

```
% make check
...
Single-process programs
-----------------------)
  alarm                       ckpt:PASSED; rstr:PASSED -> ckpt:PASSED; rstr:PASSED (2.1s)
```

Apparently, 'make check' uses a default of TEST_NAME, TAB_TO_COLUMN_30, ckpt:...
But for "Single-process programs", this causes the line to go beyond column 80.  Can we tab to column 15 instead of column 30 in the special case of "Single-process programs", or something like that, to avoid going beyond column 80?  This is more readable on a classic terminal window of 80 columns.

Probably, the same issue holds for "Thread and mutex tests" and "Process control and signals" and "Multi-process coordination:" and "IPC and shared resources" and "PTY handling" and "Interactive shells" and "Terminal-oriented programs", etc.

**Comment by karya0** [comment]
> The PR is a good one. But I have a set of comments, based also on me testing the new autotest on WSL, using Ubuntu 24.04. I use a horizontal line to separate my comments here. (And I might have had a small number of inline comments, or maybe none.)
> 
> First, two of the coderabbit nitpicks seem useful:
> 
> src/Makefile.in (1) 402-403: ⚡ Quick win
> 
> json.Po is listed for regeneration but not included in dep includes.
> 
> am__depfiles_remade now includes ./$(DEPDIR)/json.Po, but there is no matching @am__include .../json.Po include marker later in the file. This can make incremental rebuilds miss header-triggered recompiles for json.cpp.
> 
> **NOTE: I'm suggesting to use 0644, _not_ 0640. Let others read, but not write.** src/dmtcp_coordinator.cpp (1) 1941-1942: 💤 Low value
> 
> Consider restricting log file permissions.
> 
> The log file is created with mode 0666, which (before umask) allows world-readable and world-writable access. While umask typically reduces this, explicitly using more restrictive permissions like 0640 or 0644 is safer.
> 
> 🔒 Suggested fix
> 
> * ```
>    fd = open(flags.logFilename.c_str(), O_CREAT | O_WRONLY | O_APPEND, 0666);
>   ```
> 
> * ```
>    fd = open(flags.logFilename.c_str(), O_CREAT | O_WRONLY | O_APPEND, 0644);
>   ```
> 
> The `vim` test seems to be missing.
> 
> It used to be that `make check-python` would list all the tests, even though it ran only `python`. We need some way to list all the tests without running them. Maybe `make check-list` (`python3 autotest.py list`) ??
> 
> `make check 2>&1 | cat - > tmp.out` is hanging when it reaches: `Coordinator protocol tests` Doing `make check` in background should be useful. Not sure why it's hanging.
> 
> The test for command-json is failing on my Windows/WSL, running `Ubuntu 24.04.1 LTS`
> 
> command-json FAIL. exit=1 Ran 14 tests in 3.613s command: /usr/bin/python3 ./test_dmtcp_command_json.py cwd: /home/gene/dmtcp.git/test output tail (last 13 lines): ......F....... ====================================================================== FAIL: test_json_escapes_coordinator_host (**main**.DmtcpCommandJsonTest.test_json_escapes_coordinator_host) ---------------------------------------------------------------------- Traceback (most recent call last): File "/home/gene/dmtcp.git/test/./test_dmtcp_command_json.py", line 104, in test_json_escapes_coordinator_host self.assertEqual(result.returncode, 2, result.stderr) AssertionError: -11 != 2 :
> 
> ```
>   ----------------------------------------------------------------------
>   Ran 14 tests in 3.613s
> 
>   FAILED (failures=1)
> ```
> 
> Just a thought. Should we add a commit to change `DMT_` to `DMTCP_` everywhere? We had `DMT_` from the earliest days. But unusual abbreviations can be confusing. You can decide whether to add such a commit.
> 
> In `make check`, the display is going for more than 80 characters. (Although, `make check-alarm` does not go beyond 80 characters.)
> 
> ```
> % make check
> ...
> Single-process programs
> -----------------------)
>   alarm                       ckpt:PASSED; rstr:PASSED -> ckpt:PASSED; rstr:PASSED (2.1s)
> ```
> 
> Apparently, 'make check' uses a default of TEST_NAME, TAB_TO_COLUMN_30, ckpt:... But for "Single-process programs", this causes the line to go beyond column 80. Can we tab to column 15 instead of column 30 in the special case of "Single-process programs", or something like that, to avoid going beyond column 80? This is more readable on a classic terminal window of 80 columns.
> 
> Probably, the same issue holds for "Thread and mutex tests" and "Process control and signals" and "Multi-process coordination:" and "IPC and shared resources" and "PTY handling" and "Interactive shells" and "Terminal-oriented programs", etc.

Thanks for the feedback, @gc00!

Some of the coderabbit comments were already addressed in the prior commits. I have addressed the remaining comments.

DMT->DMTCP is a good idea, perhaps a separate PR after this current stream.
